File: Coin Change 2/Recursion.java
class Recursion {

    public static int coinChange(int[] coins,int index,int target) {
        if(target == 0) return 1;
        if(index < 0) return 0;

        int take = 0;
        if(coins[index] <=  target) 
            take = coinChange(coins,index,target-coins[index]);
        
        int nonTake = coinChange(coins,index-1,target);
        return nonTake + take;
    }

    public int change(int amount, int[] coins) {
        int size = coins.length;

        return coinChange(coins,size-1,amount);

    }
}

File: Coin Change 2/Recursion.py
class Solution:

    def change(self, amount: int, coins: List[int]) -> int:

        size = len(coins)
        def coinChange(coins,index,target):
            if target == 0: return 1
            if index < 0: return 0
            take = 0
            if coins[index] <= target: take = coinChange(coins,index,target-coins[index])
            nonTake = coinChange(coins,index-1,target)
            return take + nonTake

        
        return coinChange(coins,size-1,amount)